server/src/main/java/org/elasticsearch/search/lookup/SearchLookup.java

```diff
@@ -34,8 +35,11 @@
 
     final FieldsLookup fieldsLookup;
 
-    public SearchLookup(MapperService mapperService, Function<MappedFieldType, IndexFieldData<?>> fieldDataLookup) {
-        docMap = new DocLookup(mapperService, fieldDataLookup);
+    public SearchLookup(
+        MapperService mapperService,
+        BiFunction<MappedFieldType, Supplier<SearchLookup>, IndexFieldData<?>> fieldDataLookup
```

Comment:
`Supplier<SearchLookup>` is a little weird here but it is compatible with #60318 and I think it fits well with how `QueryShardContext` uses it.